### DISCUSSION TITLE
How do you report violations to devs?

### DISCUSSION THREAD

**Discussion by silasb** [Shopify/packwerk#184] [Ideas] [answered]
As we start integrating in Packwerk, we are wondering how you report violations to the dev.

Our plan was during CI we'd like to parse the reports to let the devs know there is a package/privacy violation. Even a danger plugin could work as well, but they require a machine readable report.

Has anyone needed this feature before? https://github.com/Shopify/packwerk/blob/0858c42b6cd233754127c037bb34ffeda15ad745/lib/packwerk/cli.rb#L119 seems like the critical section to export these reports as machine readable?

**Comment by exterm** [comment, marked as answer]
IIRC `packwerk check` returns machine readable output by default - should be rubocop compatible. That's the format that the VSCode extension uses.

**Reply by silasb** [reply]
The VSCode plugin uses [regular expressions](https://github.com/Gusto/packwerk-vscode/blob/main/src/outputParser.ts#L3) to handle parsing. This is something that I can work with, but ideally I'd like formatters similar to [Rubocop formatters](https://docs.rubocop.org/rubocop/formatters.html).

**Reply by exterm** [reply]
It's not super easy right now, but if you implement your own executable based on `Packwerk::Cli` you can inject a formatter https://github.com/Shopify/packwerk/blob/3119c4763d604029458ab2e18f3e479f8f71457e/lib/packwerk/cli.rb#L35

This hasn't come up a lot so far, so I'm hesitant to add a mechanism to packwerk for this. What would you like the output to look like?

**Comment by alexevanczuk** [comment]
Hey @silasb , Alex here from Gusto's Product Infrastructure team. We created the Packwerk VSCode Extension and as you said, we parse the output using regular expressions. We were also wondering if we can be able to pass in a `--json` flag to `bin/packwerk check`. As @exterm suggested, we can do this today by doing something like this:

This executes the CLI
```ruby
formatter = YourModule::YourFancyFormatter.new
style = Packwerk::OutputStyles::Coloured.new
Packwerk::Cli.new(style: style, offenses_formatter: formatter)
```

This is an implementation of `YourFancyFormatter`
```ruby
# typed: strict

# typed: strict

module YourModule
  module YourFancyFormatter
      extend T::Sig
      include Packwerk::OffensesFormatter

      sig { override.params(offenses: T::Array[T.nilable(Packwerk::Offense)]).returns(String) }
      def show_offenses(offenses)
        # Here you can do anything you want!
        # If you want, you can just go right ahead and modify the offenses output to display in a way
        # that is useful for the engineers at your company. Note that since we're dealing with in-memory
        # representations of `Packwerk::Offense`, it's inherently machine readable.
        # However, you can also serialize the array of `Packwerk::Offense` in a JSON format so that other
        # systems can parse that however they want
      end
    end
  end
end
```

If we wanted packwerk to output JSON, then we could just implement a new formatter in packwerk that is switched to with the `--json` flag. I do imagine that could be really useful to have, and perhaps the JSON could just be the same exact interface of `Packwerk::Offense` (i.e. it moves in lockstep with the currently supported public Ruby API for formatters).

I'm curious though if this solution would work for you in an interim state as we learn more about what the JSON public interface should be. If this doesn't solve things, please share what's missing 🙏

**Reply by alexevanczuk** [reply]
Hey @silasb check out https://github.com/BigRails/danger-packwerk! We just released this and I hope it can help you. Would love to hear your feedback!

**Reply by silasb** [reply]
Yessss, this looks awesome.  Will give this a try in about 2 weeks when we start migrating more code to Packwerk.

**Reply by alexevanczuk** [reply]
Sounds great! Looking forward to your feedback on it.

**Comment by alexevanczuk** [comment]
Just wanted to provide an update on this – you can now pass in your own custom offenses formatter without too much funny business!

https://github.com/Shopify/packwerk/blob/main/USAGE.md#custom-offense-formatter
